[EDNA-9]: Initialise frontend folder

pva701 opened the pull request:
[//]: # (This is a template of a pull request template.)
[//]: # (You should modify it considering specifics of a particular repository and put it there.)
[//]: # (Comments like this are meta-comments, they shouldn't be present in the final template.)
[//]: # (Comments starting with '<!---' are intended to stay in the final template.)

[//]: # (Keep in mind that it's only a template which contains items relevant to almost all conceivable repository.)
[//]: # (There can be other important items relevant to your repository that you can add here.)

## Description

- [x] Folders structure is defined
- [x] React is added as a dependency.
- [x] Webpack is used as a bundler, among with common config and config for development.
- [x] ESlint is set up (with React lint config)
- [x] Stub page is defined with svg in it (to demonstrate that svg can be imported directly into the code). Also hot reloading works as a charm.
- [x] PostCSS config with autoprefixer plugin is enabled
- [x] Link to a JS/TS styleguide added to contribution file

[//]: # (Here you can add a link to the corresponding issue tracker, e. g. https://issues.serokell.io/issue/AD-)
[//]: # (For GitHub/GitLab issues it is better to use just hash symbol or exclamation mark as it is more resistant to repo movements)
[//]: # (In this case please also prefix the link with "Resolves" keyword)
[//]: # (See https://docs.gitlab.com/ee/user/project/issues/managing_issues.html#closing-issues-automatically)
[//]: # (See https://help.github.com/en/github/managing-your-work-on-github/closing-issues-using-keywords)
## Related issue(s)

<!--
- Short description of how the PR relates to the issue, including an issue link.
For example
- Fixed #100500 by adding lenses to exported items

Write 'None' if there are no related issues (which is discouraged).
-->

## :white_check_mark: Checklist for your Pull Request

<!--
Ideally a PR has all of the checkmarks set.

If something in this list is irrelevant to your PR, you should still set this
checkmark indicating that you are sure it is dealt with (be that by irrelevance).

If you don't set a checkmark (e. g. don't add a test for new functionality),
you must be able to justify that.
-->

#### Related changes (conditional)

https://issues.serokell.io/issue/EDNA-9

#### Stylistic guide (mandatory)

- [x] My commits comply with [the policy used in Serokell](https://www.notion.so/serokell/Where-and-how-to-commit-your-work-58f8973a4b3142c8abbd2e6fd5b3a08e).

pva701 committed:
Initialise frontend folder

Problem: we don't have frontend folder yet.

Solution: initialise it with dummy project.
Use webpack qs a bundler, set up linter and other stuff.

gromakovsky commented on frontend/package.json:
```diff
@@ -0,0 +1,65 @@
+{
+  "name": "edna-app",
+  "version": "0.1.0",
+  "private": true,
```
We may make it public one day, right? At least in our dreams where this tool becomes very popular and the whole biotech community uses it…

pva701 replied:
this flag is needed to not accidentally publish it as a npm package

gromakovsky commented on frontend/src/App.tsx:
```diff
@@ -0,0 +1,15 @@
+import SerokellLogo from "~/assets/serokell.svg"
+
+export const App: FunctionComponent = (): ReactElement => {
+    // TODO remove. Code needed to check eslint react config
```
Remove when?

pva701 replied:
well, I can remove it now, I just wanted Sasha to import in it Idea/VS Code and make sure Idea highlights this, then remove. It's just crucial to have this warning turned on IDE level because sometimes you forget about some React invariants and misuse their API but their lint rules in IDE will notify you about it.

gromakovsky reviewed (changes requested):
I think we need `frontend/.gitignore` with `node_modules/` and `dist/`, right?

gromakovsky commented:
> Link to a JS/TS styleguide added to contribution file

I think it should go to `docs/code-style.md`.

pva701 commented:
>I think we need frontend/.gitignore with node_modules/ and dist/, right?

Indeed

pva701 committed:
Add TypeScript styleguide. Format all files accordingly.

gromakovsky commented on docs/web-code-style.md:
```diff
@@ -0,0 +1,20 @@
+   -->
+
+As foundation for our TypeScript/SCSS styleguide
+[this styleguide](!https://basarat.gitbook.io/typescript/styleguide) is used,
```
Why does it have this `!`? Does it work at all? Apparently it doesn't.

pva701 replied:
it worked in Idea, I will remove

gromakovsky replied:
The link works now.

gromakovsky commented on docs/web-code-style.md:
```diff
@@ -0,0 +1,20 @@
+
+As foundation for our TypeScript/SCSS styleguide
+[this styleguide](!https://basarat.gitbook.io/typescript/styleguide) is used,
+except for the followings:
```
Why do we make this exception btw?

pva701 replied:
Because we usually have every React component in a separate file (like Java class) and we usually name them in a way: `DeletionDialog`, `ModalPopup`, or something like this. We don't want to have it `deleteionDialog` and `modalPopup`

gromakovsky commented on frontend/README.md:
```diff
@@ -14,7 +14,11 @@ TODO: CI and license badges!
 ## Build
 
-`yarn build` bundles the project.
+If you are building project at the first time, you need to make `yarn install`.
```
`s/make/run` I guess

gromakovsky commented on frontend/src/index.scss:
```diff
@@ -0,0 +1 @@
+
```
Is it common to have an empty file like this one? We will add something here soon, right?

pva701 replied:
I just added it to outline file structure, to confirm it with Sasha mostly. So yes, perhaps there will be some global styles and maybe reimports of local styles, we will see.

gromakovsky commented on frontend/webpack.dev.js:
```diff
@@ -0,0 +1,27 @@
+            }
+        },
+
+        // Uncomment in case of some troubles to inspect whether dev server generates bundles
```
Uncomment what?

pva701 replied:
oh lol, writeToDisk should be commented for building speed up. Perhaps I've checked that bundles are indeed generated and forgot to comment it back

gromakovsky commented on frontend/webpack.dev.js:
```diff
@@ -0,0 +1,27 @@
+        // Redirect api requests to the backend
+        proxy: {
+            '/api': {
+                target: 'http://localhost:8080',
```
Btw, what localhost is this? Is it the machine where the server is running or the client (frontend)? By hardcoding `localhost` here do we require user to run server and browser on the same machine?

pva701 replied:
first of all, I need to notice that this is only dev config, so it's only used when you run code locally.

In production/staging frontend, backend and ngnix will be at the under the same host and this config is not be used. All api requests in the code will like /api/blabla and will be addressed to the host/IP where the website is running.

serokellcao commented on .editorconfig:
```diff
@@ -5,7 +5,7 @@
 root = true
 
-[*.hs]
+[*.{hs, ts, tsx, scss, json, js, svg}]
```
How to apply this style guide to VSCode? Will it automatically be applied? What about IntelliJ?

I'll test with VSCode now.

serokellcao replied:
VSCode works. We need to test Emacs @mkaito and IntelliJ @sashasashasasha151

pva701 replied:
This .editorconfig is most primary way to formatting nowadays. I mean it's most general (for all languages not for some specific ones) and seems to be widespread across editors.

Even Idea overrides its rules of formatting with this file. So I expect all modern editors take into consideration this file as well

serokellcao commented on frontend/src/assetTypings.d.ts:
```diff
@@ -2,22 +2,22 @@
 // without tsc type error
 
 declare module '*.svg?inline' {
```
Can you explain briefly the idea of what assetTypings is?

pva701 replied:
when you import svg files directly into tsx/ts file, tsc goes mad and start complaining about it doesn't have corresponding type declarations. So this is globally available file, where tsc will look into (as well in others in the project) to find type declarations

serokellcao commented on frontend/webpack.common.js:
```diff
@@ -3,96 +3,96 @@ const MiniCssExtractPlugin = require('mini-css-extract-plugin');
+            options: {},
+          },
+          'css-loader',
+          'postcss-loader',
```
I have this warning. Is it a big deal:

```
warning " > postcss-loader@5.0.0" has unmet peer dependency "postcss@^7.0.0 || ^8.0.1".
```

?

serokellcao replied:
Seems OK:

```
$ webpack --config webpack.dev.js
(node:7410) [DEP_WEBPACK_COMPILATION_ASSETS] DeprecationWarning: Compilation.assets will be frozen in future, all modifications are deprecated.
BREAKING CHANGE: No more changes should happen to Compilation.assets after sealing the Compilation.
        Do changes to assets earlier, e. g. in Compilation.hooks.processAssets.
        Make sure to select an appropriate stage from Compilation.PROCESS_ASSETS_STAGE_*.
asset app.4165c3a16a5e6d6707bd.js 2.47 MiB [emitted] [immutable] (name: app)
asset index.html 343 bytes [emitted]
asset style.55dda2c204c4d7d5df0a.css 1 bytes [emitted] [immutable] (name: app)
Entrypoint app 2.47 MiB = style.55dda2c204c4d7d5df0a.css 1 bytes app.4165c3a16a5e6d6707bd.js 2.47 MiB
runtime modules 670 bytes 3 modules
cacheable modules 981 KiB
  modules by path ./node_modules/ 979 KiB
    modules by path ./node_modules/scheduler/ 31.8 KiB 4 modules
    modules by path ./node_modules/react/ 70.6 KiB 2 modules
    modules by path ./node_modules/react-dom/ 875 KiB 2 modules
    ./node_modules/object-assign/index.js 2.06 KiB [built] [code generated]
  modules by path ./src/ 1.7 KiB
    modules by path ./src/*.tsx 609 bytes
      ./src/index.tsx 195 bytes [built] [code generated]
      ./src/App.tsx 414 bytes [built] [code generated]
    ./src/index.scss 50 bytes [built] [code generated]
    ./src/assets/serokell.svg 1.06 KiB [built] [code generated]
css ./node_modules/css-loader/dist/cjs.js!./node_modules/postcss-loader/dist/cjs.js!./node_modules/sass-loader/dist/cjs.js!./src/index.scss 0 bytes [code generated]
webpack 5.21.2 compiled successfully in 3814 ms
Done in 4.73s.
```

pva701 replied:
Not bit, it's causes by the fact latest PostCSS doesn't work with latest autoprefixer yet or something, so I had to downgrade PostCSS version

serokellcao reviewed (approved):
I approve.

Everything works:

  - Hot reloading
  - Formatter checked with VSCode (I hope @sashasashasasha151 will check with IntelliJ)

However, I've left some questions that I'd like to be answered in the review.

gromakovsky commented on frontend/README.md:
```diff
@@ -0,0 +1,30 @@
+## Run
+
+`yarn serve` runs the dev server.
+To operates correctly, it's needed to have backend launched on `8080` port.
```
```suggestion
To operate correctly, it's needed to have backend launched on `8080` port.
```

gromakovsky reviewed (approved):
There's one typo, otherwise LGTM. This stuff is new to me, so I can't review everything fully, but I guess I'll figure things our as we go.

Please don't merge w/o Sasha's approval though.

pva701 committed:
Add basic eslint/stylelint config with prettifier plugins

Problem: we didn't have strict rules about formatting.

Solution: Now we enabled them as eslint plugin prettifier.
Also add more eslint configs.
Also add .scss config.
Also, add more code to demonstract how it should look like
to make prettifier satisfied.

pva701 commented:
@sashasashasasha151 Perhaps, will add more rules in .eslintrc.js in future commits (future PRs)

sashasashasasha151 commented on frontend/package.json:
```diff
@@ -0,0 +1,59 @@
+  },
+  "scripts": {
+    "clean": "rm -rf dist/ && rm -rf .cache/",
+    "serve": "webpack serve --config webpack.dev.js",
```
Maybe name it `serve:dev` and add separate script for not dev serve

pva701 replied:
what is not dev:serve? I guess this one is via docker images on staging or something like this.

pva701 replied:
And this is definitely out of the scope of yarn scripts :putya:

sashasashasasha151 commented on frontend/src/.sassrc.js:
```diff
@@ -0,0 +1,10 @@
+const path = require('path');
```
Why style/lint configs are in src folder?

sashasashasasha151 commented on frontend/src/assetTypings.d.ts:
```diff
@@ -0,0 +1,23 @@
+// To allow import of assets directly from the tsx files
```
Lets move it to `src/typings`

sashasashasasha151 replied:
Style format? (just `cmd + s` or `opt + cmd + L` in IDEA)

sashasashasasha151 commented on frontend/src/assets/serokell.svg:
```diff
@@ -0,0 +1,5 @@
+<svg xmlns="http://www.w3.org/2000/svg" width="174" height="124" viewBox="0 0 62 44">
```
Lets move it to `src/assets/svg` or `src/assets/logo` (at your discretion) just to not create a dump of assets

pva701 replied:
I guess how sorting assets by its type will help you (I mean some domain separation, like in assets/MyComponent folder seems more useful to me) but I will create svg folder just to merge it quicker

sashasashasasha151 commented on frontend/src/index.tsx:
```diff
@@ -0,0 +1,9 @@
+import React from 'react';
```
Style format? (just `cmd + s` or `opt + cmd + L` in IDEA)

sashasashasasha151 commented on frontend/src/App.tsx:
```diff
@@ -0,0 +1,33 @@
+import React, { FunctionComponent, ReactElement, useEffect } from "react";
```
I think as a part of frontend init you can add `EdnaRouter` with `"/"` and `"*"` paths for main and error pages templates \
Also if we will use bem put 
```
import { withNaming } from "@bem-react/classname";

const cn = withNaming({ e: "__", m: "_", v: "_" });

export default cn;
```

to utils forder \

Also message about `normalize-scss`

sashasashasasha151 commented on frontend/package.json:
```diff
@@ -0,0 +1,59 @@
+    "webpack-dev-server": "^3.11.2",
+    "webpack-merge": "^5.7.3"
+  },
+  "scripts": {
```
Add `"check-all": "concurrently \"yarn tscompile\" \"yarn tslint\" \"yarn stylelint\""` command which will run `tscompile`, `tslint` and `stylelint` checks \
Also @gromakovsky, we need all these checks separately on CI

gromakovsky replied:
Yes, see this comment: https://issues.serokell.io/issue/EDNA-9#focus=Comments-96-21900.0-0
>  I suggest that we do EDNA-11 first and then setup CI for frontend as well.

sashasashasasha151 commented on frontend/package.json:
```diff
@@ -0,0 +1,59 @@
+    "clean": "rm -rf dist/ && rm -rf .cache/",
+    "serve": "webpack serve --config webpack.dev.js",
+    "build": "webpack --config webpack.dev.js",
+    "tscompile": "tsc",
```
```suggestion
    "tscompile": "tsc --noEmit --skipLibCheck",
```

sashasashasasha151 commented on docs/web-code-style.md:
```diff
@@ -0,0 +1,20 @@
+[this styleguide](https://basarat.gitbook.io/typescript/styleguide) is used,
+except for the followings:
+
+## Filename
```
What about forders?

pva701 replied:
what exactly? I guess `MyComponent/{MyComponent.tsx, MyComponent.scss}` structure will be used, something like this

pva701 committed:
Different improvements in frontend folder

pva701 committed:
Add @bem-react classname

It's a convenient helper to build bem-style classes.

pva701 committed:
Add normalize-scss

sashasashasasha151 committed:
Upgrade eslint config
